mlflow/store/hdfs_artifact_repo.py

```diff
@@ -0,0 +1,180 @@
+            paths = []
+            for path, is_dir, size in self._walk_path(hdfs, hdfs_base_path):
+                paths.append(FileInfo(path, is_dir, size))
+            return sorted(paths, key=lambda f: paths)
+
+    def _walk_path(self, hdfs, hdfs_path):
+        if hdfs.exists(hdfs_path) and hdfs.isdir(hdfs_path):
+            for subdir, _, files in hdfs.walk(hdfs_path):
+                yield subdir, hdfs.isdir(subdir), hdfs.info(subdir).get("size")
+                for f in files:
+                    file_path = self._join(subdir, f)
+                    yield file_path, hdfs.isdir(file_path), hdfs.info(file_path).get("size")
+
+    def download_artifacts(self, artifact_path, dst_path=None):
+        """
+            Download an artifact file or directory to a local directory if applicable, and return a
+            local path for it.
+            The caller is responsible for managing the lifecycle of the downloaded artifacts.
+
+            :param artifact_path: Relative source path to the desired artifacts.
+            :param dst_path: Absolute path of the local filesystem destination directory to which to
+                             download the specified artifacts. This directory must already exist. If
+                             unspecified, the artifacts will be downloaded to a new, uniquely-named
+                             directory on the local filesystem.
+
+            :return: Absolute path of the local filesystem location containing the downloaded
+            artifacts.
+        """
+
+        hdfs_base_path = _resolve_base_path(self.path, artifact_path)
+        local_dir = _tmp_dir(dst_path)
+
+        with hdfs_system(host=self.host, port=self.port) as hdfs:
+            for path, is_dir, _ in self._walk_path(hdfs, hdfs_base_path):
+
+                relative_path = _relative_path(hdfs_base_path, path)
+                local_path = self._join(local_dir, relative_path) if relative_path else local_dir
+
+                if is_dir:
+                    mkdir(local_path)
+                else:
+                    _download_hdfs_file(hdfs, path, local_path)
+        return local_dir
+
+    def _download_file(self, remote_file_path, local_path):
+        raise MlflowException('This is not implemented. Should never be called.')
+
+    def _join(self, base_path, path):
+        return self.get_path_module().join(base_path, path)
+
+
+@contextmanager
+def hdfs_system(host, port):
+    """
+        hdfs system context - Attempt to establish the connection to hdfs
+        and yields HadoopFileSystem
+
+    :param host: hostname or when relaying on the core-site.xml config use 'default'
+    :param port: port or when relaying on the core-site.xml config use 0
+    """
+    import pyarrow as pa
+
+    driver = os.getenv('MLFLOW_HDFS_DRIVER') or 'libhdfs'
+    kerb_ticket = os.getenv('MLFLOW_KERBEROS_TICKET_CACHE')
+    kerberos_user = os.getenv('MLFLOW_KERBEROS_USER')
```

Comment:
Do these environment variables have other standard names when used with Hadoop? It would be good to sue the existing names for these rather than creating new names for MLflow. Most users will have a shell script or something that sets these and will prefer not to modify them for MLflow.